src/charm.py

```diff
@@ -9,93 +9,377 @@
 
         # MongoDB
-        self.mongodb = DatabaseRequires(self, relation_name="mongodb", database_name=self.database)
+        self.mongodb = DatabaseRequires(
+            self,
+            relation_name=MONGODB,
+            database_name=self.database or "",
+            extra_user_roles=self.extra_user_roles or "",
+        )
         self.framework.observe(self.mongodb.on.database_created, self._on_database_created)
+        self.framework.observe(self.on[MONGODB].relation_broken, self._on_relation_broken)
 
-        self._on_config_changed(None)
+        # Kafka
+        self.kafka = KafkaRequires(
+            self,
+            relation_name=KAFKA,
+            topic=self.topic or "",
+            extra_user_roles=self.extra_user_roles or "",
+        )
+        self.framework.observe(self.kafka.on.topic_created, self._on_topic_created)
+        self.framework.observe(self.on[KAFKA].relation_broken, self._on_relation_broken)
+
+    def _on_relation_broken(self, event: RelationBrokenEvent):
+        """Handle relation broken event."""
+        if not self.unit.is_leader:
+            return
+        # update peer databag to trigger the charm status update
+        self._update_active_relations(event, "broken")
+
+    def _set_config_parameters(self, status: StatusBase):
+        """Return if the new parameters must be set."""
+        if isinstance(status, ActiveStatus):
+            return True
+        if isinstance(status, BlockedStatus) and "add relation" in status.message:
+            return True
+        return False
+
+    def _on_config_changed(self, _):
+        """Handle on config changed event."""
+        # Only execute in the unit leader
+        if not self.unit.is_leader():
+            return
+
+        # read new parameters
+        new_database_name = self.model.config.get("database-name", None)
+        new_topic = self.model.config.get("topic-name", None)
+        new_extra_user_roles = self.model.config.get("extra-user-roles", None)
+
+        # get the unit status
+        status = self._get_unit_status()
+
+        # if the status is active set new values otherwise return
+        if not self._set_config_parameters(status):
+            # set status
+            self.unit.status = status
+            return
+
+        # update parameter in the relation databag
+        self.set_secret("app", "database", new_database_name)
+        self.set_secret("app", "topic", new_topic)
+        self.set_secret("app", "extra-user-roles", new_extra_user_roles)
+
+        # update database and extra-user-roles
+        self.mysql.database = self.database or ""
+        self.mysql.extra_user_roles = self.extra_user_roles or ""
+        self.postgresql.database = self.database or ""
+        self.postgresql.extra_user_roles = self.extra_user_roles or ""
+        self.mongodb.database = self.database or ""
+        self.mongodb.extra_user_roles = self.extra_user_roles or ""
```

Comment:
If no relations are defined all this code should not really be needed (since down in the update_database_relation we have an if statement that DOES things only if the relation exists). I believe it would REALLY improve readability to have "if" statement, making clear that you do these only if a relation exists

Some thing for Kafka later on